aws/resource_aws_servicecatalog_product.go

```diff
@@ -0,0 +1,450 @@
+		// "CREATED" is not documented but seems to be the state it goes to
+		Target: []string{servicecatalog.StatusAvailable, "CREATED"},
+		Refresh: func() (interface{}, string, error) {
+			resp, err := conn.DescribeProductAsAdmin(&servicecatalog.DescribeProductAsAdminInput{
+				Id: aws.String(d.Id()),
+			})
+			if err != nil {
+				return nil, "", err
+			}
+			return resp, aws.StringValue(resp.ProductViewDetail.Status), nil
+		},
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		PollInterval: 3 * time.Second,
+	}
+	_, err := stateConf.WaitForState()
+	return err
+}
+
+func waitForServiceCatalogProductDeletion(conn *servicecatalog.ServiceCatalog, id string) error {
+	stateConf := resource.StateChangeConf{
+		Pending:      []string{servicecatalog.StatusCreating},
+		Target:       []string{""},
+		Timeout:      15 * time.Minute,
+		PollInterval: 3 * time.Second,
+		Refresh: func() (interface{}, string, error) {
+			resp, err := conn.DescribeProductAsAdmin(&servicecatalog.DescribeProductAsAdminInput{
+				Id: aws.String(id),
+			})
+			if err != nil {
+				if isAWSErr(err, servicecatalog.ErrCodeResourceNotFoundException, "") {
+					return 42, "", nil
+				}
+				return 42, "", err
+			}
+
+			return resp, aws.StringValue(resp.ProductViewDetail.Status), nil
+		},
+	}
+	_, err := stateConf.WaitForState()
+	return err
+}
+
+func resourceAwsServiceCatalogProductRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).scconn
+	log.Printf("[DEBUG] Reading Service Catalog Product with id %s", d.Id())
+	resp, err := conn.DescribeProductAsAdmin(&servicecatalog.DescribeProductAsAdminInput{
+		Id: aws.String(d.Id()),
+	})
+	if err != nil {
+		if isAWSErr(err, servicecatalog.ErrCodeResourceNotFoundException, "") {
+			log.Printf("[WARN] Service Catalog Product %q not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("reading ServiceCatalog product '%s' failed: %s", d.Id(), err)
+	}
+
+	d.Set("product_arn", resp.ProductViewDetail.ProductARN)
+
+	err = d.Set("tags", tagsToMapServiceCatalog(keyvaluetags.ServicecatalogKeyValueTags(resp.Tags).IgnoreAws().IgnoreConfig(meta.(*AWSClient).IgnoreTagsConfig).Map()))
+	if err != nil {
+		return fmt.Errorf("invalid tags read on ServiceCatalog product '%s': %s", d.Id(), err)
+	}
+
+	product := resp.ProductViewDetail.ProductViewSummary
+	d.Set("has_default_path", aws.BoolValue(product.HasDefaultPath))
+	d.Set("description", aws.StringValue(product.ShortDescription))
+	d.Set("distributor", aws.StringValue(product.Distributor))
+	d.Set("name", aws.StringValue(product.Name))
+	d.Set("owner", aws.StringValue(product.Owner))
+	d.Set("product_type", aws.StringValue(product.Type))
+	d.Set("support_description", aws.StringValue(product.SupportDescription))
+	d.Set("support_email", aws.StringValue(product.SupportEmail))
+	d.Set("support_url", aws.StringValue(product.SupportUrl))
+
+	provisioningArtifactList := make([]map[string]interface{}, 0)
```

Comment:
see earlier comment on extracting `ProvisioningArtifact` to its own function earlier and do the same for read here